consensus/src/firststep/handler.rs

```diff
@@ -106,65 +106,95 @@ impl<D: Database> MsgHandler<Message> for Reduction<D> {
         step: u8,
         committee: &Committee,
     ) -> Result<HandleMsgOutput, ConsensusError> {
+        assert_eq!(step, self.curr_step, "msg does not belong to curr step");
+
         let signature = match &msg.payload {
             Payload::Reduction(p) => Ok(p.signature),
             Payload::Empty => Ok(EMPTY_SIGNATURE),
             _ => Err(ConsensusError::InvalidMsgType),
         }?;
 
         // Collect vote, if msg payload is reduction type
-        if let Some((hash, sv)) =
+        if let Some((hash, sv, quorum_reached)) =
             self.aggr.collect_vote(committee, &msg.header, &signature)
         {
-            // Record result in global round registry of all non-Nil step_votes
-            if hash != [0u8; 32] {
-                if let Some(m) = self.sv_registry.lock().await.add_step_votes(
-                    step,
-                    hash,
-                    sv,
-                    SvType::FirstReduction,
-                ) {
-                    return Ok(HandleMsgOutput::FinalResult(m));
+            // Record result in global round registry
+            _ = self.sv_registry.lock().await.add_step_votes(
```

Comment:
`add_step_votes` return the agreement if the quorum is reached, why we discard it?

Reply:
In that particular case, it is not supposed to return any agreement. `Collect ` is called only for the current iteration, first reduction and no agreement is ready.

Keep in mind that now we have `collect_from_past` function that collects previous steps reductions and it can build an agreement at any time.